### PR TITLE
Implement PartialEq<str> for String and Cow

### PR DESCRIPTION
**Pull request by gkoz**
Right now comparing a `&String` (or a `&Cow`) to a `&str` requires redundant borrowing of the latter. Implementing `PartialEq<str>` tries to avoid this limitation.

``` rust
struct Foo (String);

fn main () {
    let s = Foo("foo".to_string());
    match s {
        Foo(ref x) if x == &"foo" => println!("foo!"),
        // avoid this -----^
        _ => {}
    }
}
```

I was hoping that #23521 would solve this but it didn't work out.

### CODE REVIEW

File: src/libcollections/string.rs [hunk 1]
```diff
@@ -826,6 +826,38 @@ impl_eq! { String, &'a str }
 impl_eq! { Cow<'a, str>, String }
```
Comment (petrochenkov) [inline src/libcollections/string.rs]: Btw, you can use `&[..]` notation inside of `impl_eq!` like #23521 does and reduce all the impls for `Cow<'a, str>`, `String` and `str` to just a few lines.

Comment (gkoz) [inline src/libcollections/string.rs]: Thanks for the tip!

**Comment by gkoz** [comment]
I've put up a less verbose version of the patch in accordance with @petrochenkov's advice.

**Comment by alexcrichton** [comment]
@bors: r+ 5234063

Thanks!

**Pull request merged**